File: api/app/schemas/due_date_override.py
"""Model Due Date Override schemas."""
from pydantic import BaseModel, ConfigDict, Field, field_validator
from datetime import datetime, date
from typing import Optional, Literal
from app.schemas.user import UserResponse


class DueDateOverrideCreate(BaseModel):
    """Schema for creating a due date override."""
    override_type: Literal["ONE_TIME", "PERMANENT"] = Field(
        ..., description="ONE_TIME auto-clears after validation, PERMANENT persists and rolls forward"
    )
    target_scope: Literal["CURRENT_REQUEST", "NEXT_CYCLE"] = Field(
        ..., description="CURRENT_REQUEST affects active validation, NEXT_CYCLE affects future"
    )
    override_date: date = Field(
        ..., description="New due date (must be earlier than calculated and in the future)"
    )
    reason: str = Field(
        ..., min_length=10, description="Justification for override (min 10 characters)"
    )

    @field_validator('reason')
    @classmethod
    def validate_reason_length(cls, v: str) -> str:
        if len(v.strip()) < 10:
            raise ValueError("Reason must be at least 10 characters")
        return v.strip()


class DueDateOverrideClear(BaseModel):
    """Schema for clearing/canceling an override."""
    reason: str = Field(
        ..., min_length=10, description="Reason for clearing the override"
    )

    @field_validator('reason')
    @classmethod
    def validate_reason_length(cls, v: str) -> str:
        if len(v.strip()) < 10:
            raise ValueError("Reason must be at least 10 characters")
        return v.strip()


class DueDateOverrideResponse(BaseModel):
    """Response schema for a single override."""
    override_id: int
    model_id: int
    validation_request_id: Optional[int] = None
    override_type: str
    target_scope: str
    override_date: date
    original_calculated_date: date
    reason: str
    created_by_user_id: int
    created_by_user: UserResponse
    created_at: datetime
    is_active: bool
    cleared_at: Optional[datetime] = None
    cleared_by_user_id: Optional[int] = None
    cleared_by_user: Optional[UserResponse] = None
    cleared_reason: Optional[str] = None
    cleared_type: Optional[str] = None
    superseded_by_override_id: Optional[int] = None
    rolled_from_override_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class CurrentDueDateOverrideResponse(BaseModel):
    """Response schema for current override status."""
    model_id: int
    model_name: str
    has_active_override: bool
    active_override: Optional[DueDateOverrideResponse] = None

    # Calculated dates for reference
    policy_calculated_date: Optional[date] = None
    effective_due_date: Optional[date] = None  # min(override, calculated) - enforces earlier-only

    # Validation request context (if applicable)
    current_validation_request_id: Optional[int] = None
    current_validation_status: Optional[str] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class DueDateOverrideHistoryResponse(BaseModel):
    """Response schema for override history."""
    model_id: int
    model_name: str
    active_override: Optional[DueDateOverrideResponse] = None
    override_history: list[DueDateOverrideResponse] = []

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

File: web/src/api/dueDateOverride.ts
/**
 * Due Date Override API
 *
 * Allows admins to override a model's validation submission due date
 * to an earlier date (accelerate validation schedule).
 */
import api from './client';

export interface UserResponse {
    user_id: number;
    email: string;
    full_name: string;
    role: string;
}

export interface DueDateOverride {
    override_id: number;
    model_id: number;
    validation_request_id: number | null;
    override_type: 'ONE_TIME' | 'PERMANENT';
    target_scope: 'CURRENT_REQUEST' | 'NEXT_CYCLE';
    override_date: string;
    original_calculated_date: string;
    reason: string;
    created_by_user_id: number;
    created_by_user: UserResponse;
    created_at: string;
    is_active: boolean;
    cleared_at: string | null;
    cleared_by_user_id: number | null;
    cleared_by_user: UserResponse | null;
    cleared_reason: string | null;
    cleared_type: 'MANUAL' | 'AUTO_VALIDATION_COMPLETE' | 'AUTO_ROLL_FORWARD' | 'AUTO_REQUEST_CANCELLED' | 'SUPERSEDED' | null;
    superseded_by_override_id: number | null;
    rolled_from_override_id: number | null;
}

export interface CurrentDueDateOverrideResponse {
    model_id: number;
    model_name: string;
    has_active_override: boolean;
    active_override: DueDateOverride | null;
    policy_calculated_date: string | null;
    effective_due_date: string | null;
    current_validation_request_id: number | null;
    current_validation_status: string | null;
}

export interface DueDateOverrideHistoryResponse {
    model_id: number;
    model_name: string;
    active_override: DueDateOverride | null;
    override_history: DueDateOverride[];
}

export interface CreateDueDateOverrideRequest {
    override_type: 'ONE_TIME' | 'PERMANENT';
    target_scope: 'CURRENT_REQUEST' | 'NEXT_CYCLE';
    override_date: string;
    reason: string;
}

export interface ClearDueDateOverrideRequest {
    reason: string;
}

export const dueDateOverrideApi = {
    /**
     * Get current due date override status for a model
     */
    getForModel: async (modelId: number): Promise<CurrentDueDateOverrideResponse> => {
        const response = await api.get(`/models/${modelId}/due-date-override`);
        return response.data;
    },

    /**
     * Create a due date override (Admin only)
     */
    create: async (modelId: number, data: CreateDueDateOverrideRequest): Promise<DueDateOverride> => {
        const response = await api.post(`/models/${modelId}/due-date-override`, data);
        return response.data;
    },

    /**
     * Clear an active due date override (Admin only)
     */
    clear: async (modelId: number, data: ClearDueDateOverrideRequest): Promise<DueDateOverride> => {
        const response = await api.delete(`/models/${modelId}/due-date-override`, { data });
        return response.data;
    },

    /**
     * Get full history of due date overrides for a model
     */
    getHistory: async (modelId: number): Promise<DueDateOverrideHistoryResponse> => {
        const response = await api.get(`/models/${modelId}/due-date-override/history`);
        return response.data;
    }
};

export default dueDateOverrideApi;
